Galaxy in AnVIL Demo Chapter (#28)
fhdsl/AnVIL_Demos: 05-fastqc-galaxy-in-anvil.Rmd

```{r echo = FALSE}
knitr::opts_chunk$set(out.width = "100%")
```

```{r echo=FALSE}
library(cow)
```

# (PART\*) Run a Tool With Galaxy {-}   

# Overview {#run-tool-with-galaxy-overview}

This chapter contains instructions for the SARS-CoV-2 sequencing data FastQC analysis with Galaxy in AnVIL activity.

> Galaxy is a free computing platform that allows users to upload or "fetch" data from online repositories and analyze that data with supported tools using a point and click user interface (UI) rather than with a terminal or other API. Galaxy in AnVIL provides a cloud instance of Galaxy within AnVIL such that the benefits of AnVIL (such as secure data sharing, access to controlled access data) are combined with the benefits of running a reproducible analysis with Galaxy. However, Galaxy in AnVIL is not free. 

Reviewer: ```suggestion
Galaxy is a free computing platform that allows users to upload or "fetch" data from online repositories and analyze that data with supported tools using a [point and click user interface](https://en.wikipedia.org/wiki/Point_and_click) (UI) rather than with a [command-line interface](https://en.wikipedia.org/wiki/Command-line_interface). Galaxy in AnVIL means that the benefits of AnVIL (such as secure data sharing, access to controlled access data) are combined with the benefits of running a reproducible analysis with Galaxy. 

Like all compute on AnVIL, Galaxy in AnVIL will incur costs. 
```

> This activity will 

Reviewer: empty line?

Author: Thanks for catching!! I will fill in.

Author: Realized I had the sentence I was going to write at the beginning so removed the incomplete sentence

## Learning Objectives

This activity will teach you how to use the AnVIL platform to:

1. Launch Galaxy in AnVIL

> 1. Import data into Galaxy in AnVIL

Reviewer: Typically we say "Galaxy on AnVIL" not in AnVIL, so let's update to be consistent

Author: Should be fixed in [commit 7f16341](https://github.com/fhdsl/AnVIL_Demos/pull/28/commits/7f163419fc481db153290a77975dc54ee4c56a6f)

> 1. Examine sequence data files (fastq files)

Reviewer: ```suggestion
1. Examine sequence data files (`.fastq` files)
```

> 1. Understand how to find, open, and run a tool (FastQC)

Reviewer: ```suggestion
1. Find, open, and run a tool (FastQC)
```

1. Examine the quality control summary report (Webpage)

> 1. Learn how to export your results for future viewing

Reviewer: ```suggestion
1. Export your results for future viewing
```

> 1. Show how to shut down Galaxy in AnVIL

Reviewer: ```suggestion
1. Shut down Galaxy in AnVIL
```

# Getting Started {#run-tool-with-galaxy-getting-started}

## Account Login

If you don't have an AnVIL account, you will need to create one using the steps below in the AnVIL Account Creation Dropdown section.

> <details><summary>AnVIL Account Creation</summary>

Reviewer: Yay 😄

## Workspace Set Up

Workspaces are the building blocks of projects in Terra. Inside a Workspace, you can run analyses, launch interactive tools like RStudio and Galaxy, store data, and share results. 

Cloning an existing Workspace allows you to copy existing documentation, code, and/or data into your own experimental space. The data for this activity is already available on AnVIL, specifically the [`demos-galaxy-fastqc` workspace](https://anvil.terra.bio/#workspaces/anvil-outreach/demos-galaxy-fastqc), so you'll want to clone the repository. 

When cloning, AnVIL makes a copy of notebooks and code for you to modify. Data however, is linked back to the original Workspace through Data Tables, which saves space!

::: {.notice}
The screenshots included below show a different workspace than what this tutorial will be using. For this tutorial, you want to clone the [`demos-galaxy-fastqc` workspace](https://anvil.terra.bio/#workspaces/anvil-outreach/demos-galaxy-fastqc) 

Be sure to name the cloned workspace a unique name in step 6. We recommend adding `_yourfirstname_yourlastname` to the workspace name.
:::

> :::: {.borrowed_chunk}

Reviewer: This version allows you to set the workspace name, etc.
https://github.com/jhudsl/AnVIL_Template/blob/main/08-student_modules.Rmd#L37-L51

Author: Used that version -- love it!

```{r, echo = FALSE, results='asis'}
cow::borrow_chapter(
  doc_path = "child/_child_workspace_clone.Rmd",
  repo_name = "jhudsl/AnVIL_Template"
)
```
::::

# Activity {#run-tool-with-galaxy-activity}

## Launching Galaxy in AnVIL

:::: {.borrowed_chunk}
```{r, echo = FALSE, results='asis'}
cow::borrow_chapter(
  doc_path = "child/_child_galaxy_start.Rmd",
  repo_name = "jhudsl/AnVIL_Template"
)
```
::::

### Navigating Galaxy in AnVIL

> 

Reviewer: love the borrowed chunks!!!

:::: {.borrowed_chunk}
```{r, echo=FALSE, results='asis'}
cow::borrow_chapter(
  doc_path = "child/_child_galaxy_navigate.Rmd",
  repo_name = "jhudsl/AnVIL_Template"
)
```
::::

## Importing Data into Galaxy in AnVIL

> When we cloned our workspace, our cloned workspace linked to the original data in the workspace! We will upload three files from the AnVIL workspace into Galaxy, though we need only one `fastq` data sequence file for our activity. The others will be used if you want to continue with a related activity that performs alignment and variant discovery after quality control. These three files are (1) the forward and (2) the reverse reads for our sample, as well as, (3) the reference genome for SARS-CoV-2. There are two sets of reads for our sample because the scientists who collected it used paired-end sequencing. The sample files we are looking at end in `fastq` because they are raw data from the sequencer. The reference genome ends in `.fasta` because it has already been cleaned up by scientists. 

Reviewer: ```suggestion
When we cloned our workspace, our cloned workspace linked to the original data! We will upload three files from the AnVIL workspace into Galaxy, though we need only one `fastq` data sequence file for our activity. The others will be used if you want to continue with a related activity that performs alignment and variant discovery after quality control. These three files are (1) the forward and (2) the reverse reads for our sample, as well as, (3) the reference genome for SARS-CoV-2. There are two sets of reads for our sample because the scientists who collected it used paired-end sequencing. The sample files we are looking at end in `fastq` because they are raw data from the sequencer. The reference genome ends in `.fasta` because it has already been cleaned up by scientists. 
```

Might also link to the wikipedia page on paired end sequencing

```

2. Click on “Choose remote files” at the bottom of the popup.

```{r, echo=FALSE, fig.alt='Screenshot of the Galaxy Data upload popup page and upload options.'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zre8qeE9x56RSjusIF3XcuHoxqGf30nfdvjwCoEMd8o/edit#slide=id.g325209e7a14_0_1112")
```

::: {.notice}
If you had files locally on your computer that you wanted to upload, you would use the "Choose local file" button.

Or if you had files you wanted to import from a data repository like Zenodo, you would use the "Paste/Fetch data" button.

We're using the "Choose remote files" button because we have data in our AnVIL workspace that we can import into the Galaxy in AnVIL instance.
:::

3. Double-click the Workspace folder.

```{r, echo=FALSE, fig.alt="Screenshot of the Galaxy Data upload popup pane, highlighting the AnVIL workspace where your data is linked"}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zre8qeE9x56RSjusIF3XcuHoxqGf30nfdvjwCoEMd8o/edit#slide=id.g325209e7a14_0_1117")
```

4. Upload the sample sequence data files

>   4A. Double-click “Tables/”

Reviewer: Note this looks a little funky. I'd just remove the 4A, 4B headers.
![image](https://github.com/user-attachments/assets/5af18327-7db5-47d0-9815-f1948e4ff4a5)

Author: @avahoffman would you still use (A) or A or should I use `-` or just have it be lines under the 4 heading?

Author: Used bullet points. I like how it looks

```{r, echo=FALSE, fig.alt="Screenshot of the Galaxy Data upload popup pane, highlighting the Tables/ folder where linked data will be in a cloned workspace"}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zre8qeE9x56RSjusIF3XcuHoxqGf30nfdvjwCoEMd8o/edit#slide=id.g325209e7a14_0_1122")
```

  4B. Then double-click “sample/”.

```{r, echo=FALSE, fig.alt="Screenshot of the Galaxy Data upload popup pane, highlighting the sample/ folder where the data sequence .fastq files will be found"}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zre8qeE9x56RSjusIF3XcuHoxqGf30nfdvjwCoEMd8o/edit#slide=id.g325209e7a14_0_1126")
```

  4C. Click the two sample `.fastq` file checkboxes to select them.

```{r, echo=FALSE, fig.alt="Screenshot of the Galaxy Data upload popup pane, selecting the two data sequence .fastq files"}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zre8qeE9x56RSjusIF3XcuHoxqGf30nfdvjwCoEMd8o/edit#slide=id.g325209e7a14_0_1132")
```

>   4D. These files will be highlighted in green and click “Ok”.

Reviewer: ```suggestion
  4D. These files will be highlighted in green when ready. Click “Ok”.
```

ottrpal::include_slide("https://docs.google.com/presentation/d/1zre8qeE9x56RSjusIF3XcuHoxqGf30nfdvjwCoEMd8o/edit#slide=id.g325209e7a14_0_1122")
```

  6B. This time, double click "reference/".

```{r, echo = FALSE, fig.alt="Screenshot of the Galaxy Data upload popup pane, highlighting the reference/ folder where the reference genome fasta file will be found."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zre8qeE9x56RSjusIF3XcuHoxqGf30nfdvjwCoEMd8o/edit#slide=id.g3257f1d5e0e_0_56")
```

  6C. Click the `fasta` file.

```{r, echo=FALSE, fig.alt = "Screenshot of the Galaxy Data upload popup pane, highlighting the reference genome .fasta file"}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zre8qeE9x56RSjusIF3XcuHoxqGf30nfdvjwCoEMd8o/edit#slide=id.g325209e7a14_0_1148")
```

  6D. This file will be highlighted green and click "Ok".

```{r, echo=FALSE, fig.alt = "Screenshot of the Galaxy Data upload popup pane, highlighting the selected reference genome .fasta file and the Ok button."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zre8qeE9x56RSjusIF3XcuHoxqGf30nfdvjwCoEMd8o/edit#slide=id.g325209e7a14_0_1153")
```

</details>

> 

Reviewer: Add some <br> here before step 7? 
![image](https://github.com/user-attachments/assets/2de5e401-a192-49f6-bdc6-99c090e4c4b0)

Author: Sorry @avahoffman  I'm not sure I understand this comment/suggestion. Providing a little of context of what I did below.

I made this a details section for two reasons. One it's not required for this tutorial, unless you want to do next steps (like alignment). Two, wondering if it would be helpful for people to try to do the steps themselves since the process is super similar to the previous steps.

Author: I tried to clarify that they should expand that sections for Steps 5-6 (and added "Expand for" for all details sections). Also added a line break for more space

Reviewer: Sorry it interpreted `<br>` as an actual breakline 🤦🏻‍♀️ 

I just thought these two lines were kinda close together

Author: Oh gotcha! That makes more sense :) I'm happy I didn't fully get it though, I think it's clearer now

```

<details><summary>FastQC summary report explained</summary>

For more information on the [contents of the output quality control summary report from FastQC, consider this resource from Michigan State](https://rtsf.natsci.msu.edu/genomics/technical-documents/fastqc-tutorial-and-faq.aspx)

::: {.reflection}
QUESTIONS:

1. Explore “Basic Statistics”. How many total reads are there? Have any been flagged as poor quality? What is the sequence length?

2. Explore “Per base sequence quality”. Based on the Basic Statistics, is 28-40 a good or bad quality score? 

3. Is it okay to proceed based on the per base sequence quality?
:::

</details>

## Exporting your results

In case you want to view the results later, you can download the file. 

1.  Click on the name of the results you want to export/save, and it will expand the info shown for that file. Click the floppy disk/save icon.

```{r, echo=FALSE, fig.alt = "Screenshot of Galaxy in AnVIL highlighting the floppy disk/save icon for the results that you want to export."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zre8qeE9x56RSjusIF3XcuHoxqGf30nfdvjwCoEMd8o/edit#slide=id.g3257f1d5e0e_0_107")
```

> This will download a zip (compressed) file with the results to your computer. Uncompress it to view it locally

Reviewer: ```suggestion
This will download a zip (compressed) file with the results to your computer. Uncompress it to view it locally.
```

## Shutting down Galaxy in AnVIL

:::: {.borrowed_chunk}
```{r, echo = FALSE, results='asis'}
cow::borrow_chapter(
  doc_path = "child/_child_galaxy_delete.Rmd",
  repo_name = "jhudsl/AnVIL_Template"
)
```
::::

# Next steps {#run-tool-with-galaxy-next-steps}

> This tutorial is just a first step in working with this data, specifically considering the quality scores of one set of the sequencing reads. Next steps include running FastQC on the reverse reads and downstream applications such as alignment and variant detection. Walk throughs for these next steps can be found in this [Genomic Data Science Community Network Lab Exercise Book](https://jhudatascience.org/GDSCN_Book_SARS_Galaxy_on_AnVIL/student-activity-guide.html#alignment) or [this AnVIL Demos Recording](https://www.youtube.com/watch?v=_uT0IWL1wso)

Reviewer: ```suggestion
This tutorial is just a first step in working with this data, specifically considering the quality scores of one set of the sequencing reads. Next steps include running FastQC on the reverse reads and downstream applications such as alignment and variant detection. Walk throughs for these next steps can be found in this [Genomic Data Science Community Network Lab Exercise Book](https://jhudatascience.org/GDSCN_Book_SARS_Galaxy_on_AnVIL/student-activity-guide.html#alignment) or [this AnVIL Demos Recording](https://www.youtube.com/watch?v=_uT0IWL1wso).
```
